hbase-server/src/main/java/org/apache/hadoop/hbase/master/normalizer/SimpleRegionNormalizer.java

```diff
@@ -371,7 +358,11 @@ private boolean skipForMerge(final RegionStates regionStates, final RegionInfo r
       final long nextSizeMb = getRegionSizeMB(next);
       // always merge away empty regions when they present themselves.
       if (currentSizeMb == 0 || nextSizeMb == 0 || currentSizeMb + nextSizeMb < avgRegionSizeMb) {
-        plans.add(new MergeNormalizationPlan(current, next));
+        final MergeNormalizationPlan plan = new MergeNormalizationPlan.Builder()
+          .addTarget(current, currentSizeMb)
+          .addTarget(next, nextSizeMb)
```

Comment:
Yes, merging more than two regions at a time is coming in [HBASE-24419](https://issues.apache.org/jira/browse/HBASE-24419). I started working on it yesterday. And yes, this constructor API for MergeNormalizationPlan will likely change because of it.

Reply:
This PR is not posted, #2490.

Reply:
> Regardless, maybe we want to validate `list_size() % 2 == 0` in `MergeNormalizationPlan` constructor which initializes `normalizationTargets`?

I still think that this validation is better suitable for this current PR because we are introducing at least the flexibility to submit variable no of target regions for merging.

Reply:
See the logic of `MergeTableRegionsProcedure#checkRegionsToMerge`, that's where validation of any merge request is applied. I think it's not the normalizer's job to verify these parameters. And anyway, the user isn't specifying the regions, they only select a table. Normalizer's algorithm generates the region merge list.

Reply:
Sure this is not normalizer's job to validate but by keeping target regions as list, we might keep odd no of regions a while just to abandon them later by `MergeTableRegionsProcedure#checkRegionsToMerge` if we keep it unbounded list.
How about `List<Pair<NormalizerRegionInfo, NormalizerRegionInfo>>` as target regions in `MergeNormalizationPlan`? 
`NormalizerRegionInfo` can just contain `regionInfo` and `regionSize`.

Reply:
I haven't rebased it yet, but have a look at 450c4d5 on #2490 . The limit to pairs is removed by that patch, so I'll just have to undo the pair-wise API you propose.

I could remove the use of the list from this patch, go back to using the first and second member variables, if you think that aspect is so critical. My push-back is because, in my opinion, it's not the POJO's responsibility to ensure these invariants, it's the job of the merge system. And the merge system already makes these checks, so let's leave that logic written in one place, and leave it close to where it's consumed.

Reply:
I don't have strong opinion for API change since merge system is already taking care of validation.

You might prefer adding a small javadoc for `Builder.addTarget()` explaining to add even no of regions with the sequence in which client wants to merge them?

Reply:
Sure, let me buff up the javadoc on that class. FYI, technically an odd number of regions can be merged, so long a they're more than 2 and in a contiguous sequence.

Reply:
I extended the class-level javadoc to remind callers about the expectations of the `mergeRegions` method.